includes/abstracts/abstract-wc-stripe-payment-gateway.php

```diff
@@ -717,6 +721,54 @@ public function prepare_order_source( $order = null ) {
 		);
 	}
 
+	/**
+	 * Retrieves a source based on an order. Checks for the customer, associated
+	 * with the order and/or the customer which was manually entered in the admin.
+	 *
+	 * @param WC_Order $order The order to use as a starting point.
+	 * @return stdClass|null  Either an object with the necessary data or null.
+	 */
+	public function get_default_customer_source_for_order( $order ) {
+		$stripe_customer = null;
+		$order_id        = WC_Stripe_Helper::is_wc_lt( '3.0' ) ? $order->id : $order->get_id();
+
+		// Check for explicit manual input first.
+		$stripe_customer_id = get_post_meta( $order_id, '_stripe_customer_id', true );
+		if ( ! empty( $stripe_customer_id ) ) {
+			$stripe_customer = new WC_Stripe_Customer();
+			$stripe_customer->set_id( $stripe_customer_id );
+		}
+
+		// If there is a WooCommerce customer to work with, check if they have a Stripe ID.
+		if ( empty( $stripe_customer ) ) {
+			$wc_customer = $order->get_user();
+			if ( $wc_customer ) {
+				$stripe_customer = new WC_Stripe_Customer( $wc_customer->ID );
+				if ( empty( $stripe_customer->get_id() ) ) {
+					$stripe_customer = null;
+				}
+			}
+		}
+
+		// No customer to work with, neither coming from WooCommerce nor Subscription's fields.
+		if ( empty( $stripe_customer ) ) {
+			return null;
+		}
+
+		// Try to retrieve the default source of the customer.
+		$source_object = $stripe_customer->get_default_source();
```

Comment:
If we can't retrieve the default source, can't we still make the renewal payment, relying on only sending the `customer` without `source`?

Reply:
Why would that happen?

The main reason for this PR and the issue(s) behind it is the fact that Stripe said we shouldn't send customers without specifying a source, and that this behavior is not supported and will probably break in the future.

Additionally, by getting the `default_source` from the customer object, we should already be covering the default behavior of Stripe when no source is passed.

Reply:
I think the main issue here is _why_ Stripe is asking us to send a `source` with every PaymentIntents request. They're asking us to send a source because in the future, the concept of `default source` is going away.

This seems to imply that if we aren't able to make a PaymentIntents request without a source parameter, we _also_ won't be able to make an API call to retrieve the default source if we don't already know it.

Meaning, I don't think this API call will work:

```
WC_Stripe_API::request( array(), 'sources/' . $default_token->get_token(), 'GET' );
```

The reason we wanted to save the source we get back from the PaymentIntent request is because we know that many users have already saved their subscriptions without specifying a `source`, and we want to add a `source`. That way there are less subscriptions out there without a `source` in the future, when `default_source` is deprecated.

I followed up with Stripe about this to double-check my reasoning.